grype/version/pep440_constraint_test.go

```diff
@@ -0,0 +1,305 @@
+			version:    "17.12.0-ce-rc1.0.20200309214505-aa6a9891b09c+incompatible",
+			constraint: "< v1.5",
+			satisfied:  false,
+		},
+		{
+			name:       "strip unbalanced v from right side >",
+			version:    "17.12.0-ce-rc1.0.20200309214505-aa6a9891b09c+incompatible",
+			constraint: "> v1.5",
+			satisfied:  true,
+		},
+		{
+			name:       "rc candidates with no '-' can match semver pattern",
+			version:    "1.20rc1",
+			constraint: " = 1.20.0-rc1",
+			satisfied:  true,
+		},
+		{
+			name:       "candidates ahead of alpha",
+			version:    "3.11.0",
+			constraint: "> 3.11.0-alpha1",
+			satisfied:  true,
+		},
+		{
+			name:       "candidates ahead of beta",
+			version:    "3.11.0",
+			constraint: "> 3.11.0-beta1",
+			satisfied:  true,
+		},
+		{
+			name:       "candidates ahead of same alpha versions",
+			version:    "3.11.0-alpha5",
+			constraint: "> 3.11.0-alpha1",
+			satisfied:  true,
+		},
+		{
+			name:       "candidates are placed correctly between alpha and release",
+			version:    "3.11.0-beta5",
+			constraint: "3.11.0 || = 3.11.0-alpha1",
+			satisfied:  false,
+		},
+		{
+			name:       "candidates with letter suffix are alphabetically greater than their versions",
+			version:    "1.0.2a",
+			constraint: " < 1.0.2w",
+			satisfied:  true,
+		},
+		{
+			name:       "candidates with multiple letter suffix are alphabetically greater than their versions",
+			version:    "1.0.2zg",
+			constraint: " < 1.0.2zh",
+			satisfied:  true,
+		},
+		{
+			name:       "candidates with pre suffix are sorted numerically",
+			version:    "1.0.2pre1",
+			constraint: " < 1.0.2pre2",
+			satisfied:  true,
+		},
+		{
+			name:       "candidates with letter suffix and r0 are alphabetically greater than their versions",
+			version:    "1.0.2k-r0",
+			constraint: " < 1.0.2l-r0",
+			satisfied:  true,
+		},
+		{
+			name:       "openssl version with letter suffix and r0 are alphabetically greater than their versions",
+			version:    "1.0.2k-r0",
+			constraint: ">= 1.0.2",
+			satisfied:  true,
+		},
+		{
+			name:       "openssl versions with letter suffix and r0 are alphabetically greater than their versions and compared equally to other lettered versions",
+			version:    "1.0.2k-r0",
+			constraint: ">= 1.0.2, < 1.0.2m",
+			satisfied:  true,
+		},
+		{
+			name:       "openssl pre2 is still considered less than release",
+			version:    "1.1.1-pre2",
+			constraint: "> 1.1.1-pre1, < 1.1.1",
+			satisfied:  true,
+		},
+		{
+			name:       "major version releases are less than their subsequent patch releases with letter suffixes",
+			version:    "1.1.1",
+			constraint: "> 1.1.1-a",
+			satisfied:  true,
+		},
+		{
+			name:       "date based pep440 version string boundary condition",
+			version:    "2022.12.7",
+			constraint: ">=2017.11.05,<2022.12.07",
+		},
+	}
+	for _, tc := range tests {
+		t.Run(tc.name, func(t *testing.T) {
+			_, err := newPep440Constraint(tc.constraint)
+			require.NoError(t, err)
```

Comment:
Big `TODO` here - didn't actually implement these tests yet. Test cases are copied from the old fuzzy matcher.